### ISSUE TITLE
MonoGame DesktopGL does not respect SamplerState.PointClamp in custom shader

### ISSUE THREAD

**Issue by vchelaru** [MonoGame/MonoGame#6393]
Setting `GraphicsDevice.SamplerStates[0] = SamplerState.PointClamp;` results in point filtering applied in XNA 4.0 PC, but the same code/shader result in linear filtering in Desktop GL.

To reproduce the bug:

1. Download https://drive.google.com/file/d/1U-JU5iEKw-fEAXOLDuqDxcjxIDhvTKkU/view?usp=sharing
1. Unzip the file to a Windows 10 machine (XNA will be required to compare XNA behavior)
1. Open XnaCompare.sln (located in the root of the unzipped folder)
1. Run the project
1. Observe that the rendered quad has point filtering (hard edges between pixels) ![image](https://user-images.githubusercontent.com/729631/43369165-49a6a04e-9326-11e8-9c5a-5ae43963d7c0.png)
1. Open DesktopGlShaderTest/DesktopGlShaderTest.sln
1. Run the project
1. Observe that the rendered quad has linear filtering (smoothed pixels) ![image](https://user-images.githubusercontent.com/729631/43369175-84e560f0-9326-11e8-949c-dfcf3ae2b652.png)

Note: Both projects use the same shader file, but the shader file is added directly as an XNB file to the DesktopGL project (due to a missing feature of linked files in the content builder). I did try building the shader with the latest dev build of the content pipeline tool - I don't know if the bug is caused by the pipeline tool or MonoGame runtime.

This was tested on a Windows 10 machine with the following hardware:
 * i7-8550U CPU @ 1.80GHz
 * 16.0 GB ram
 * 64 bit OS
 * NVIDIA GeForce 940MX

#### What version of MonoGame does the bug occur on:
- MonoGame 3.7.0.1657

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:

- DesktopGL, compared against Microsoft XNA 4.0

**Comment by cra0zy** [comment]
Turns out, some other stuff leak from `SpriteBatch` besides `PointClamp`, yikes.

**Comment by vchelaru** [comment]
For what it's worth, in this case I'm not using SpriteBatch anywhere in the project, so...maybe not the same bug?

**Comment by Jjagg** [comment]
@cra0zy SpriteBatch is not supposed to reset graphics state after ending a batch. It doesn't do that in XNA either.

**Comment by cra0zy** [comment]
> @cra0zy SpriteBatch is not supposed to reset graphics state after ending a batch. It doesn't do that in XNA either.

Interesting, I never realized that... well at least that explains why the "workaround" is working.

**Comment by Jjagg** [comment]
I couldn't reproduce the issue. What's likely happening here is the sampler is assigned to another slot than what's expected.

That said the order of samplers sometimes gets messed up when you have multiple of them in an effect. I don't think that issue is actually logged so I'll open a new one.

**Comment by Jjagg** [comment]
Nope, I was wrong. The issue is actually that we override the sampler state with whatever sampler state is set in the effect file (if any). While XNA only overrides the properties that are set from the effect file, we also override the others with default values.

**Comment by vchelaru** [comment]
Thanks for looking into that @Jjagg  . I was concerned when you said you couldn't reproduce it, as I have a linked project that reproduces it for me.

**Comment by Jjagg** [comment]
I tested first with a new project which did not set a sampler state in the effect file.

**Comment by mattj1** [comment]
I know this is late, but I had a similar issue and I seemed to fix it by rendering everything to a separate target first.
